### PR TITLE
set cpu affinity to numa in MULTI situation

### PR DESCRIPTION
**Pull request by wgzintel**
### Details:
- *Hold threads for GPU for MULTI:GPU,CPU*
- *MULTI set CPU affinity to “NUMA” during load network*

### Tickets:
 - *ticket-89110*

### CODE REVIEW

File: src/plugins/auto/plugin.cpp [hunk 1]
```diff
@@ -495,39 +495,98 @@ IExecutableNetworkInternal::Ptr MultiDeviceInferencePlugin::LoadNetworkImpl(cons
...
+        if (deviceName.find("CPU") != std::string::npos) {
+            key_streamsNums = PluginConfigParams::KEY_CPU_THROUGHPUT_STREAMS;
+            key_threadsNums = PluginConfigParams::KEY_CPU_THREADS_NUM;
+        } else if (deviceName.find("GPU") != std::string::npos) {
+            key_streamsNums = PluginConfigParams::KEY_GPU_THROUGHPUT_STREAMS;
+            key_threadsNums = GPUConfigParams::KEY_GPU_MAX_NUM_THREADS;
+        }
+        // print CPU or GPU streams num and threads num
+        if (!key_streamsNums.empty() && !key_threadsNums.empty()) {
+            std::string sStreamNums = exec_net->GetConfig(key_streamsNums).as<std::string>();
```
Comment (songbell) [inline src/plugins/auto/plugin.cpp:536]: better put in try catch block

File: src/plugins/auto/plugin.cpp [hunk 2]
```diff
@@ -495,39 +495,98 @@ IExecutableNetworkInternal::Ptr MultiDeviceInferencePlugin::LoadNetworkImpl(cons
...
+        loads.clear();
+    }
+
+    // Finally load the CPU
+    if (cpuDevice.size() > 0) {
+        if (!executableNetworkPerDevice.empty()) {
+            // If the other devices load successfully, then set NUMA to CPU
+            GetCore()->set_property(cpuDevice[0].deviceName, ov::affinity(ov::Affinity::NUMA));
+        }
+        priorityDevices.emplace_back(cpuDevice[0]);
```
Comment (songbell) [inline src/plugins/auto/plugin.cpp:582]: with this change, the original device priority order will not be respected. Is this the expected logic?

File: src/plugins/auto/plugin.cpp [hunk 3]
```diff
@@ -496,39 +496,103 @@ IExecutableNetworkInternal::Ptr MultiDeviceInferencePlugin::LoadNetworkImpl(cons
...
     std::vector<Task> loads;
     std::once_flag readNetworkFlag;
-    for (auto& p : metaDevices) {
-        loads.push_back([&]() {
-            auto tmpiter = fullConfig.find(CONFIG_KEY(ALLOW_AUTO_BATCHING));
-            if (tmpiter != fullConfig.end()) {
-                if (tmpiter->second == PluginConfigParams::NO)
-                    multiSContext->_batchingDisabled = true;
-                p.config.insert({tmpiter->first, tmpiter->second});
+    std::vector<DeviceInformation> cpuDevice;
```
Comment (songbell) [inline src/plugins/auto/plugin.cpp:499]: do we have situation when more than one CPU device? maybe if not, we do not need vector here, as below many logic only use the first element in this vector.

Comment (wgzintel) [reply src/plugins/auto/plugin.cpp]: Modified to not use vector for CPU

**Comment by wangleis** [comment]
Please disable "enable_hyper_thread" when CPU plugin enable it as private key

File: src/plugins/auto/plugin.cpp [hunk 4]
```diff
@@ -496,40 +496,102 @@ IExecutableNetworkInternal::Ptr MultiDeviceInferencePlugin::LoadNetworkImpl(cons
...
+        loads.clear();
+    }
+
+    // Finally load the CPU
+    if (iterCPU != metaDevices.end()) {
+        if (!executableNetworkPerDevice.empty() && iterCPU->config.find(ov::affinity.name()) == iterCPU->config.end()) {
+            LOG_DEBUG_TAG("set affinity to NUMA for CPU");
+            // If the other devices load successfully and no user set affinity then set NUMA to CPU
+            iterCPU->config.insert({ov::affinity.name(), ov::affinity(ov::Affinity::NUMA).second.as<std::string>()});
+            iterCPU->config.insert({CONFIG_KEY_INTERNAL(ENABLE_HYPER_THREAD), CONFIG_VALUE_INTERNAL(YES)});
```
Comment (peterchen-intel) [inline src/plugins/auto/plugin.cpp:586]: CONFIG_VALUE_INTERNAL(NO)

File: src/plugins/auto/plugin.cpp [hunk 5]
```diff
@@ -496,40 +496,102 @@ IExecutableNetworkInternal::Ptr MultiDeviceInferencePlugin::LoadNetworkImpl(cons
...
+            key_threadsNums = PluginConfigParams::KEY_CPU_THREADS_NUM;
+        } else if (deviceName.find("GPU") != std::string::npos) {
+            key_streamsNums = PluginConfigParams::KEY_GPU_THROUGHPUT_STREAMS;
+            key_threadsNums = GPUConfigParams::KEY_GPU_MAX_NUM_THREADS;
+        }
+        // print CPU or GPU streams num and threads num
+        if (!key_streamsNums.empty() && !key_threadsNums.empty()) {
+            try {
+                std::string sStreamNums = exec_net->GetConfig(key_streamsNums).as<std::string>();
+                std::string sThreadNums = exec_net->GetConfig(key_threadsNums).as<std::string>();
```
Comment (sunxiaoxia2022) [inline src/plugins/auto/plugin.cpp:539]: Change GetConfig() to GetMetric() to get inference_num_threads.

File: src/inference/src/threading/ie_istreams_executor.cpp [hunk 6]
```diff
@@ -410,18 +424,19 @@ IStreamsExecutor::Config IStreamsExecutor::Config::MakeDefaultMultiThreaded(cons
...
-                             //    latency case, single-node yet hybrid case that uses
-                             //      all core types
-                             //      or
-                             //      big-cores only, but the #cores is "enough" (pls see the logic above)
-                             // it is usually beneficial not to use the hyper-threading (which is default)
-                             : num_cores_default;
+    const auto hwCores =
+        !bLatencyCase && numaNodesNum == 1
+            // throughput case on a single-NUMA node machine uses all available cores
+            ? (!streamExecutorConfig._enable_hyper_thread ? num_cores_default : parallel_get_max_threads())
```
Comment (wangleis) [inline src/inference/src/threading/ie_istreams_executor.cpp:430]: Please use `(streamExecutorConfig._enable_hyper_thread ? parallel_get_max_threads() : num_cores_default)` to align with previous `(enable_hyper_thread ? parallel_get_max_threads() : getNumberOfCPUCores())`

Comment (wgzintel) [reply src/inference/src/threading/ie_istreams_executor.cpp]: done

**Review by wangleis — APPROVED** [review]

File: src/inference/dev_api/cpp_interfaces/interface/ie_internal_plugin_config.hpp [hunk 7]
```diff
@@ -104,6 +104,17 @@ DECLARE_CONFIG_KEY(FORCE_DISABLE_CACHE);
...
+ * @brief enable hyper thread
+ */
+DECLARE_CONFIG_KEY(ENABLE_HYPER_THREAD);
+
+/**
+ * @brief generic boolean values
+ */
+DECLARE_CONFIG_VALUE(YES);
+DECLARE_CONFIG_VALUE(NO);
+
```
Comment (songbell) [inline src/inference/dev_api/cpp_interfaces/interface/ie_internal_plugin_config.hpp:117]: I suppose config value not needed to re-declare?

File: src/inference/src/threading/ie_istreams_executor.cpp [hunk 8]
```diff
@@ -280,6 +282,16 @@ void IStreamsExecutor::Config::SetConfig(const std::string& key, const std::stri
                        << ". Expected only non negative numbers";
         }
         _small_core_offset = val_i;
+    } else if (key == CONFIG_KEY_INTERNAL(ENABLE_HYPER_THREAD)) {
+        int val_i;
+        if (value == CONFIG_VALUE_INTERNAL(YES)) {
+            val_i = true;
```
Comment (songbell) [inline src/inference/src/threading/ie_istreams_executor.cpp:288]: why not directly assign value to _enable_hyper_thread

File: src/plugins/auto/plugin.cpp [hunk 9]
```diff
@@ -496,40 +496,100 @@ IExecutableNetworkInternal::Ptr MultiDeviceInferencePlugin::LoadNetworkImpl(cons
...
+        executor->runAndWait(loads);
+        loads.clear();
+    }
+
+    // Finally load the CPU
+    if (iterCPU != metaDevices.end()) {
+        if (!executableNetworkPerDevice.empty() && iterCPU->config.find(ov::affinity.name()) == iterCPU->config.end()) {
+            LOG_DEBUG_TAG("set affinity to NUMA and disable hyper thread for CPU");
+            // If the other devices load successfully and no user set affinity then set NUMA to CPU
+            iterCPU->config.insert({ov::affinity.name(), ov::affinity(ov::Affinity::NUMA).second.as<std::string>()});
```
Comment (songbell) [inline src/plugins/auto/plugin.cpp:583]: probably , if app loadnetwork with ov::affinity already, your insert may not take effect

File: src/inference/src/threading/ie_istreams_executor.cpp (hunk 8 above)
Comment (sunxiaoxia2022) [reply src/inference/src/threading/ie_istreams_executor.cpp:288]: Ok, I will remove the redundant code.

File: src/inference/dev_api/cpp_interfaces/interface/ie_internal_plugin_config.hpp (hunk 7 above)
Comment (wgzintel) [reply src/inference/dev_api/cpp_interfaces/interface/ie_internal_plugin_config.hpp]: done

File: src/inference/src/threading/ie_istreams_executor.cpp (hunk 8 above)
Comment (wgzintel) [reply src/inference/src/threading/ie_istreams_executor.cpp]: done

**Review by dmitry-gorokhov — APPROVED** [review]

**Pull request merged**